isy-security/src/main/java/de/bund/bva/isyfact/security/oauth2/client/IsyOAuth2Authentifizierungsmanager.java

```diff
@@ -2,89 +2,150 @@
+    }
+
+    @Override
+    public void authentifiziereSystem(String issuerLocation, String clientId, String clientSecret, String username, String password, @Nullable String bhknz)
             throws AuthenticationException {
-        Assert.notNull(registrationId, "registrationId cannot be null");
+        Assert.notNull(issuerLocation, "issuerLocation cannot be null");
+        Assert.notNull(clientId, "clientId cannot be null");
+        Assert.notNull(clientSecret, "clientSecret cannot be null");
         Assert.notNull(username, "username cannot be null");
         Assert.notNull(password, "password cannot be null");
 
-        PasswordAuthenticationToken token = new PasswordAuthenticationToken(registrationId, username, password, bhknz);
+        ClientRegistration clientRegistration = ClientRegistrations.fromIssuerLocation(issuerLocation)
+                .clientId(clientId)
+                .clientSecret(clientSecret)
+                .authorizationGrantType(AuthorizationGrantType.PASSWORD)
+                .build();
 
-        authenticateAndChangeAuthenticatedPrincipal(token);
+        Authentication unauthenticatedToken = new PasswordClientRegistrationAuthenticationToken(clientRegistration, username, password, bhknz);
+        authenticateAndChangeAuthenticatedPrincipal(unauthenticatedToken);
+    }
+
+    /**
+     * Creates an appropriate authentication token for the authorization grant type configured for the registration ID.
+     *
+     * @param oauth2ClientRegistrationId
+     *         registration ID to create the token for
+     * @return an unauthenticated token that can be passed to the provider manager
+     */
+    private Authentication getAuthenticationTokenForRegistrationId(String oauth2ClientRegistrationId) {
+        ClientRegistration clientRegistration = null;
+        if (clientRegistrationRepository != null) {
+            clientRegistration = clientRegistrationRepository.findByRegistrationId(oauth2ClientRegistrationId);
+        }
+        Assert.notNull(clientRegistration, "Could not find ClientRegistration with id '" + oauth2ClientRegistrationId + "'");
+
+        // load additional props for this registration ID, can be null
+        AdditionalRegistrationProperties props = isyOAuth2ClientProps.getRegistration().get(clientRegistration.getRegistrationId());
+        String bhknz = null;
+        if (props != null) {
+            // the BHKNZ is optional but can be set for CC or ROPC
+            bhknz = props.getBhknz();
+        }
+
+        AuthorizationGrantType grantType = clientRegistration.getAuthorizationGrantType();
+        if (AuthorizationGrantType.CLIENT_CREDENTIALS.equals(grantType)) {
+            return new ClientCredentialsRegistrationIdAuthenticationToken(oauth2ClientRegistrationId, bhknz);
+        } else if (AuthorizationGrantType.PASSWORD.equals(grantType)) {
+            // ROPC requires the username and password to be set in the additional properties
+            if (props == null) {
```

Comment:
That's correct, but as you can see the line 141 with  `if(props == null)` leads in case of true to `throw new BadCredentialsException` with the error message `No configured credentials (username, password) found for client with ...`. 

But this will only occur if no `isy.security`-property is set at all. If only the property `bhknz` is set, but the `username` and `password` are still missing, this error-check will not take effect. 

On my opinion the correct if-statement for that error-message should be:
`if(props == null && props.username == null && props.password==null)`

this statement checks all possible constellations for exactly that error-message.

Reply:
if expression will be adapted